flaky test: `TestChallengerCountersPreGenesisGame`

**Test:** `TestChallengerCountersPreGenesisGame`
**Job:** `memory-all-opn-op-reth`
**Package:** `github.com/ethereum-optimism/optimism/op-acceptance-tests/tests/interop/proofs`
**CI run:** https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/125637/workflows/ad4fd185-ce73-4b8a-a65d-0b5915b7d40a/jobs/5070898
**Date:** 2026-05-19
**Branch:** `gh-readonly-queue/develop/pr-19847-2fff789c7e40a4631ac01dfaff1a66da09cb89a1` (merge queue staging for #19847)

## Failure

```
Error Trace:  op-devstack/sysgo/pre_genesis_super_game.go:129
              op-devstack/sysgo/multichain_supernode_runtime.go:225
              op-devstack/sysgo/multichain_proofs.go:153
              op-devstack/presets/interop.go:97
              op-acceptance-tests/tests/interop/proofs/challenger_test.go:90
Error:        "6" is not less than "6"
Messages:     pre-genesis dispute game must be created before the chosen rollup start block
--- FAIL: TestChallengerCountersPreGenesisGame (40.03s)
```

Setup logs from the same run show `planned_rollup_start_block=6`, so the assertion `disputeGameBlock < plannedRollupStartBlock` failed with both values equal to 6. Looks like a boundary/race condition in pre-genesis super-game setup where the dispute-game-creation tx lands in the same L1 block that is later selected as the rollup start block.

## Flake history

200-pipeline scan on `develop` shows 187 consecutive passes since the last failure in window, 0.5% worst-case fail rate, P(streak)=36.9% — "Possibly fixed" but with low confidence; today's failure confirms it is still live.

## Reproduction hint

Stress the test under load to hit the L1 block-progression race:

```
cd op-acceptance-tests
go test ./tests/interop/proofs -run '^TestChallengerCountersPreGenesisGame$' -count=50 -race
```

Investigate `op-devstack/sysgo/pre_genesis_super_game.go:129` and how `planned_rollup_start_block` is computed relative to the L1 block in which the game-creation tx is mined.

--- a/op-devstack/sysgo/pre_genesis_super_game.go
+++ b/op-devstack/sysgo/pre_genesis_super_game.go
@@ -26,24 +26,12 @@ import (
 // current head at which the rollup's first L1 block (and L2 genesis) is
 // scheduled. The dispute game creation transaction must be included in an
 // L1 block strictly before this target. The delay is the wall-clock budget
-// (in L1 blocks of l1Net.blockTime each) for:
-//   - the OPCM migration call that registers the super game type and
-//     commits the starting anchor, and
-//   - the dispute game creation submission and inclusion.
-//
-// At an L1 block time of 6s, a delay of 20 gives ~120s of budget — well
-// beyond observed setup times (~40s worst-case on CI). The post-migration
-// head check below converts any remaining budget exhaustion into a fast,
-// explicit setup failure rather than a late-stage flaky assertion.
-const preGenesisRollupStartBlockDelay = uint64(20)
-
-// preGenesisCreateBlockMargin is the minimum number of L1 blocks that must
-// remain between the head observed immediately before submitting the
-// dispute game creation transaction and the planned rollup start block.
-// One block accounts for the standard mining latency between submission
-// and inclusion; the extra margin absorbs an additional block of jitter
-// (e.g. an L1 block being mined while the tx is still in the mempool).
-const preGenesisCreateBlockMargin = uint64(2)
+// (in L1 blocks of l1Net.blockTime each) for the OPCM migration call and
+// the dispute game creation submission and inclusion. At an L1 block time
+// of 6s, a delay of 10 gives ~60s of budget — headroom above the ~40s
+// worst-case observed on CI without making the test wait unnecessarily long
+// for the planned start block.
+const preGenesisRollupStartBlockDelay = uint64(10)
 
 var preGenesisStartingAnchorRoot = common.HexToHash("0x0000000000000000000000000000000000000000000000000000000000000042")
 
@@ -137,30 +125,6 @@ func preparePreGenesisSuperGame(
 	)
 	initBond, err := contractio.Read(dgf.InitBonds(uint32(gameType)), t.Ctx())
 	require.NoError(err, "failed to read dispute game init bond")
-
-	// Read the L1 head immediately before submitting the dispute game
-	// creation tx. The migration above mines a variable number of L1
-	// blocks; if it has consumed so much of the block budget that
-	// inclusion of the create tx would collide with (or pass) the planned
-	// rollup start block, fail the setup explicitly here instead of
-	// letting the assertion below fire intermittently. This converts the
-	// underlying wall-clock race (anchor pre-commitment locks plannedGenesisTime
-	// and thus plannedRollupStartBlockNumber, while the create tx is included
-	// at the L1 head reached at submission time) into a clear, debuggable
-	// setup failure.
-	preCreateHeader, err := client.HeaderByNumber(t.Ctx(), nil)
-	require.NoError(err, "failed to fetch L1 head before dispute game creation")
-	preCreateHead := bigs.Uint64Strict(preCreateHeader.Number)
-	t.Logger().Info("post-migration L1 head before dispute game creation",
-		"l1_head", preCreateHead,
-		"planned_rollup_start_block", plannedRollupStartBlockNumber,
-		"margin", preGenesisCreateBlockMargin,
-	)
-	require.Greaterf(plannedRollupStartBlockNumber, preCreateHead+preGenesisCreateBlockMargin,
-		"L1 setup race: head=%d, planned rollup start=%d, margin=%d. "+
-			"Setup consumed too many L1 blocks before dispute game creation. "+
-			"Increase preGenesisRollupStartBlockDelay or reduce setup work.",
-		preCreateHead, plannedRollupStartBlockNumber, preGenesisCreateBlockMargin)
 
 	receipt, err := contractio.Write(
 		dgf.Create(uint32(gameType), rootClaim, extraData),

--- a/op-devstack/sysgo/pre_genesis_super_game.go
+++ b/op-devstack/sysgo/pre_genesis_super_game.go
@@ -22,7 +22,28 @@ import (
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
-const preGenesisRollupStartBlockDelay = uint64(6)
+// preGenesisRollupStartBlockDelay is the number of L1 blocks ahead of the
+// current head at which the rollup's first L1 block (and L2 genesis) is
+// scheduled. The dispute game creation transaction must be included in an
+// L1 block strictly before this target. The delay is the wall-clock budget
+// (in L1 blocks of l1Net.blockTime each) for:
+//   - the OPCM migration call that registers the super game type and
+//     commits the starting anchor, and
+//   - the dispute game creation submission and inclusion.
+//
+// At an L1 block time of 6s, a delay of 20 gives ~120s of budget — well
+// beyond observed setup times (~40s worst-case on CI). The post-migration
+// head check below converts any remaining budget exhaustion into a fast,
+// explicit setup failure rather than a late-stage flaky assertion.
+const preGenesisRollupStartBlockDelay = uint64(20)
+
+// preGenesisCreateBlockMargin is the minimum number of L1 blocks that must
+// remain between the head observed immediately before submitting the
+// dispute game creation transaction and the planned rollup start block.
+// One block accounts for the standard mining latency between submission
+// and inclusion; the extra margin absorbs an additional block of jitter
+// (e.g. an L1 block being mined while the tx is still in the mempool).
+const preGenesisCreateBlockMargin = uint64(2)
 
 var preGenesisStartingAnchorRoot = common.HexToHash("0x0000000000000000000000000000000000000000000000000000000000000042")
 
@@ -116,6 +137,31 @@ func preparePreGenesisSuperGame(
 	)
 	initBond, err := contractio.Read(dgf.InitBonds(uint32(gameType)), t.Ctx())
 	require.NoError(err, "failed to read dispute game init bond")
+
+	// Read the L1 head immediately before submitting the dispute game
+	// creation tx. The migration above mines a variable number of L1
+	// blocks; if it has consumed so much of the block budget that
+	// inclusion of the create tx would collide with (or pass) the planned
+	// rollup start block, fail the setup explicitly here instead of
+	// letting the assertion below fire intermittently. This converts the
+	// underlying wall-clock race (anchor pre-commitment locks plannedGenesisTime
+	// and thus plannedRollupStartBlockNumber, while the create tx is included
+	// at the L1 head reached at submission time) into a clear, debuggable
+	// setup failure.
+	preCreateHeader, err := client.HeaderByNumber(t.Ctx(), nil)
+	require.NoError(err, "failed to fetch L1 head before dispute game creation")
+	preCreateHead := bigs.Uint64Strict(preCreateHeader.Number)
+	t.Logger().Info("post-migration L1 head before dispute game creation",
+		"l1_head", preCreateHead,
+		"planned_rollup_start_block", plannedRollupStartBlockNumber,
+		"margin", preGenesisCreateBlockMargin,
+	)
+	require.Greaterf(plannedRollupStartBlockNumber, preCreateHead+preGenesisCreateBlockMargin,
+		"L1 setup race: head=%d, planned rollup start=%d, margin=%d. "+
+			"Setup consumed too many L1 blocks before dispute game creation. "+
+			"Increase preGenesisRollupStartBlockDelay or reduce setup work.",
+		preCreateHead, plannedRollupStartBlockNumber, preGenesisCreateBlockMargin)
+
 	receipt, err := contractio.Write(
 		dgf.Create(uint32(gameType), rootClaim, extraData),
 		t.Ctx(),

--- a/op-devstack/sysgo/pre_genesis_super_game.go
+++ b/op-devstack/sysgo/pre_genesis_super_game.go
@@ -22,15 +22,8 @@ import (
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
-// preGenesisRollupStartBlockDelay is the number of L1 blocks ahead of the
-// current head at which the rollup's first L1 block (and L2 genesis) is
-// scheduled. The dispute game creation transaction must be included in an
-// L1 block strictly before this target. The delay is the wall-clock budget
-// (in L1 blocks of l1Net.blockTime each) for the OPCM migration call and
-// the dispute game creation submission and inclusion. At an L1 block time
-// of 6s, a delay of 10 gives ~60s of budget — headroom above the ~40s
-// worst-case observed on CI without making the test wait unnecessarily long
-// for the planned start block.
+// Budget (in L1 blocks) for migration + dispute game creation. At blockTime=6s,
+// 10 gives ~60s, above the ~40s worst case observed on CI. See #20869.
 const preGenesisRollupStartBlockDelay = uint64(10)
 
 var preGenesisStartingAnchorRoot = common.HexToHash("0x0000000000000000000000000000000000000000000000000000000000000042")

